File: src/odioctl/upgrade/check.py
"""`odioctl upgrade check` — compare state.json against the published manifest
and refresh /var/cache/odio/upgrades.json (wired to a daily systemd user timer)."""

from __future__ import annotations

import argparse
import contextlib
import json
import os
import sys
from dataclasses import dataclass
from datetime import UTC, datetime
from typing import TypedDict, cast

from odioctl import components, manifest, state
from odioctl.manifest import Manifest
from odioctl.state import SYSTEM_STATE_PATH, SYSTEM_UPGRADES_PATH, State
from odioctl.versions import parse_version


class RoleUpgrade(TypedDict):
    """One per-role entry in the upgrades.json `roles` list."""

    name: str
    installed: str
    available: str


class UpgradeReport(TypedDict):
    """Schema of upgrades.json (written by `check`, read by `apply` and odio-api).

    `roles` is a delta — only roles whose target > installed appear, in the
    {name, installed, available} shape consumed by odio-motd. `manifest` is
    the full target snapshot cached so `apply` can skip the network round-
    trip once `check` has run. `pending_components` lists what the user
    enabled but is not installed yet ("role:x" / "feature:y"); they make
    `upgrade_available` true on their own so odio-ui offers the upgrade and
    `apply` does not refuse it.

    `latest` is the version the target release calls itself; `target_tag` is
    the GitHub tag it is published under, which `apply` needs to build its
    install.sh URL. They are the same string for a normal release and differ
    for a pre-release ("2026.7.0rc2-9-gcad916c" published as "pr-84").
    """

    current: str
    latest: str
    target_tag: str
    upgrade_available: bool
    roles: list[RoleUpgrade]
    pending_components: list[str]
    manifest: Manifest
    checked_at: str


@dataclass
class CheckOptions:
    state: str = SYSTEM_STATE_PATH
    version: str | None = None
    output: str = SYSTEM_UPGRADES_PATH


def add_check_arguments(p: argparse.ArgumentParser) -> None:
    p.description = "Compare local state against the remote manifest and refresh upgrades.json."
    p.add_argument("--state", default=SYSTEM_STATE_PATH)
    p.add_argument(
        "--version",
        help="release tag to compare against, for a box running a pre-release "
        f"(default: the published latest, or ${manifest.ODIOS_VERSION_ENV})",
    )
    p.add_argument("--output", default=SYSTEM_UPGRADES_PATH)


def check_from_args(ns: argparse.Namespace) -> int:
    return run_check(CheckOptions(state=ns.state, version=ns.version, output=ns.output))


def _compute_role_upgrades(st: State, man: Manifest) -> list[RoleUpgrade]:
    upgrades: list[RoleUpgrade] = []
    for role, installed in st["roles"].items():
        # No version = enabled here, not installed yet: that's a pending
        # component, not a role upgrade (components.REQUESTED_VERSION).
        if not installed:
            continue
        available = man["roles"].get(role)
        if available and parse_version(available) > parse_version(installed):
            upgrades.append({"name": role, "installed": installed, "available": available})
    upgrades.sort(key=lambda r: r["name"])
    return upgrades


def _build_upgrades_report(
    st: State, man: Manifest, target_tag: str | None = None
) -> UpgradeReport:
    upgrades = _compute_role_upgrades(st, man)
    pending = components.pending_components(st, set(man["roles"]))
    current = st["odios"]
    latest = man["odios"]
    newer = parse_version(latest) > parse_version(current)
    return {
        "current": current,
        "latest": latest,
        "target_tag": target_tag or latest,
        "upgrade_available": bool(upgrades) or newer or bool(pending),
        "roles": upgrades,
        "pending_components": pending,
        "manifest": man,
        "checked_at": datetime.now(UTC).strftime("%Y-%m-%dT%H:%M:%SZ"),
    }


def _write_upgrades_report(report: UpgradeReport, output: str) -> None:
    os.makedirs(os.path.dirname(output), exist_ok=True)
    with open(output, "w") as f:
        json.dump(report, f, indent=2)
        f.write("\n")
    # Default umask gives 0644; explicit chmod so other `odio` group members
    # (target_user when the timer runs as them, ansible become_user, etc.)
    # can rewrite this file without needing root.
    with contextlib.suppress(OSError):
        os.chmod(output, 0o664)


def _print_check_summary(report: UpgradeReport) -> None:
    if report["target_tag"] != report["latest"]:
        print(f"Comparing against release {report['target_tag']} ({report['latest']})")
    if report["upgrade_available"]:
        print(f"Upgrades available: {report['current']} → {report['latest']}")
        for r in report["roles"]:
            print(f"  {r['name']}: {r['installed']} → {r['available']}")
        for c in report["pending_components"]:
            print(f"  {c}: pending install")
    else:
        print(f"Up to date ({report['current']})")


def read_report(upgrades_path: str) -> UpgradeReport | None:
    """The cached upgrades.json, or None when missing/unreadable."""
    try:
        with open(upgrades_path) as f:
            data = json.load(f)
    except (OSError, json.JSONDecodeError):
        return None
    if not isinstance(data, dict) or "manifest" not in data:
        return None
    data.setdefault("pending_components", [])
    data.setdefault("target_tag", data.get("latest") or "latest")
    return cast(UpgradeReport, data)


def refresh(opts: CheckOptions) -> UpgradeReport | None:
    """Refresh upgrades.json after a local change (component toggle).

    Fetches the manifest like `run_check`; when that fails (offline) the cached
    manifest in the existing upgrades.json is reused so `pending_components`
    and `upgrade_available` still reflect the new state. None when neither a
    manifest nor a cache is available (nothing was written).
    """
    try:
        st = state.read_state(opts.state)
    except (OSError, json.JSONDecodeError, state.StateError):
        return None
    try:
        url, tag = manifest.check_source(opts.version)
    except ValueError:
        return None
    man = manifest.fetch_manifest(url)
    if man is None:
        cached = read_report(opts.output)
        if cached is None:
            return None
        man = cached["manifest"]
        # Reusing that manifest means staying on the release it came from:
        # dropping its tag here would leave `apply` building an install.sh URL
        # from a version string that is not a tag.
        tag = tag or cached["target_tag"]
    report = _build_upgrades_report(st, man, tag)
    try:
        _write_upgrades_report(report, opts.output)
    except OSError:
        return None
    return report


def run_check(opts: CheckOptions) -> int:
    try:
        st = state.read_state(opts.state)
    except (OSError, json.JSONDecodeError, state.StateError) as e:
        print(f"Error reading state: {e}", file=sys.stderr)
        return 2

    try:
        url, tag = manifest.check_source(opts.version)
    except ValueError as e:
        print(f"Error: {e}", file=sys.stderr)
        return 2

    man = manifest.fetch_manifest(url)
    if man is None:
        return 2

    report = _build_upgrades_report(st, man, tag)
    _write_upgrades_report(report, opts.output)
    _print_check_summary(report)
    return 1 if report["upgrade_available"] else 0

File: upgrade/check.go
// Package upgrade implements `odioctl upgrade check|apply|verify`: the
// upgrades.json contract with odio-ui, and the install.sh re-run.
package upgrade

import (
	"bytes"
	"cmp"
	"encoding/json"
	"fmt"
	"io"
	"log/slog"
	"os"
	"path/filepath"
	"slices"
	"time"

	"github.com/b0bbywan/odioctl/components"
	"github.com/b0bbywan/odioctl/fsutil"
	"github.com/b0bbywan/odioctl/manifest"
	"github.com/b0bbywan/odioctl/state"
	"github.com/b0bbywan/odioctl/versions"
)

// RoleUpgrade is one per-role entry of the upgrades.json delta, in the exact
// {name, installed, available} shape odio-motd reads.
type RoleUpgrade struct {
	Name      string `json:"name"`
	Installed string `json:"installed"`
	Available string `json:"available"`
}

// Report is the schema of upgrades.json (written by `check`, read by `apply`
// and odio-api). Roles is a delta, Manifest the snapshot that spares `apply`
// the network, Latest the version and TargetTag the release it comes from.
type Report struct {
	Current           string            `json:"current"`
	Latest            string            `json:"latest"`
	TargetTag         string            `json:"target_tag"`
	UpgradeAvailable  bool              `json:"upgrade_available"`
	Roles             []RoleUpgrade     `json:"roles"`
	PendingComponents []string          `json:"pending_components"`
	Manifest          manifest.Manifest `json:"manifest"`
	CheckedAt         string            `json:"checked_at"`
}

type CheckOptions struct {
	State   string // default state.SystemStatePath
	Version string // release tag override, "" = published latest / env
	Output  string // default state.UpgradesPathFor(State)
}

func (o CheckOptions) withDefaults() CheckOptions {
	if o.State == "" {
		o.State = state.SystemStatePath
	}
	if o.Output == "" {
		o.Output = state.UpgradesPathFor(o.State)
	}
	return o
}

func computeRoleUpgrades(st state.State, man manifest.Manifest) []RoleUpgrade {
	upgrades := []RoleUpgrade{}
	for role, installed := range st.Roles {
		// No version = enabled here, not installed yet: that's a pending
		// component, not a role upgrade (components.RequestedVersion).
		if installed == "" {
			continue
		}
		available := man.Roles[role]
		if available != "" && versions.Compare(available, installed) > 0 {
			upgrades = append(upgrades, RoleUpgrade{Name: role, Installed: installed, Available: available})
		}
	}
	slices.SortFunc(upgrades, func(a, b RoleUpgrade) int {
		return cmp.Compare(a.Name, b.Name)
	})
	return upgrades
}

func buildReport(st state.State, man manifest.Manifest, targetTag string) Report {
	upgrades := computeRoleUpgrades(st, man)
	pending := components.Pending(st, man.Roles)
	if pending == nil {
		pending = []string{}
	}
	if targetTag == "" {
		targetTag = man.Odios
	}
	return Report{
		Current:           st.Odios,
		Latest:            man.Odios,
		TargetTag:         targetTag,
		UpgradeAvailable:  len(upgrades) > 0 || versions.Compare(man.Odios, st.Odios) > 0 || len(pending) > 0,
		Roles:             upgrades,
		PendingComponents: pending,
		Manifest:          man,
		CheckedAt:         time.Now().UTC().Format("2006-01-02T15:04:05Z"),
	}
}

func writeReport(report Report, output string) error {
	if dir := filepath.Dir(output); dir != "." {
		if err := os.MkdirAll(dir, 0o755); err != nil {
			return err
		}
	}
	var buf bytes.Buffer
	enc := json.NewEncoder(&buf)
	enc.SetEscapeHTML(false)
	enc.SetIndent("", "  ")
	if err := enc.Encode(report); err != nil {
		return err
	}
	// Atomic: the timer and every web toggle rewrite this while `apply` reads
	// it, and a torn read silently falls back to "latest".
	if err := fsutil.AtomicWriteText(output, buf.String()); err != nil {
		return err
	}
	// 0664 so other `odio` group members (target_user under the timer,
	// ansible become_user) can rewrite it without root.
	_ = os.Chmod(output, 0o664)
	return nil
}

func printCheckSummary(w io.Writer, r Report) {
	if r.TargetTag != r.Latest {
		fmt.Fprintf(w, "Comparing against release %s (%s)\n", r.TargetTag, r.Latest)
	}
	if !r.UpgradeAvailable {
		fmt.Fprintf(w, "Up to date (%s)\n", r.Current)
		return
	}
	fmt.Fprintf(w, "Upgrades available: %s → %s\n", r.Current, r.Latest)
	for _, ru := range r.Roles {
		fmt.Fprintf(w, "  %s: %s → %s\n", ru.Name, ru.Installed, ru.Available)
	}
	for _, c := range r.PendingComponents {
		fmt.Fprintf(w, "  %s: pending install\n", c)
	}
}

// HasPending reports whether ref ("role:x" / "feature:y") is pending.
func (r *Report) HasPending(ref string) bool {
	return slices.Contains(r.PendingComponents, ref)
}

// ReadReport returns the cached upgrades.json, nil when missing, unreadable
// or not a report `check` wrote (no manifest snapshot, no target tag).
func ReadReport(path string) *Report {
	b, err := os.ReadFile(path)
	if err != nil {
		return nil
	}
	var probe struct {
		Manifest json.RawMessage `json:"manifest"`
	}
	if json.Unmarshal(b, &probe) != nil || probe.Manifest == nil {
		return nil
	}
	var r Report
	if json.Unmarshal(b, &r) != nil || r.TargetTag == "" {
		return nil
	}
	if r.PendingComponents == nil {
		r.PendingComponents = []string{}
	}
	return &r
}

// check is the pipeline behind RunCheck and Refresh: state.json against the
// target manifest, written to upgrades.json. RunCheck always fetches; Refresh
// reuses the cached manifest, never under another tag than the one it was
// written for — `apply` trusts target_tag to name the manifest next to it.
func check(opts CheckOptions, useCache bool) (*Report, error) {
	opts = opts.withDefaults()
	st, err := state.Read(opts.State)
	if err != nil {
		return nil, fmt.Errorf("reading state: %w", err)
	}
	url, tag, err := manifest.CheckSource(opts.Version)
	if err != nil {
		return nil, err
	}
	var man *manifest.Manifest
	if useCache {
		if cached := ReadReport(opts.Output); cached != nil && (tag == "" || cached.TargetTag == tag) {
			man = &cached.Manifest
			tag = cached.TargetTag
		}
	}
	if man == nil {
		if man, err = manifest.Fetch(url); err != nil {
			return nil, fmt.Errorf("could not fetch manifest at %s: %w", url, err)
		}
	}
	report := buildReport(st, *man, tag)
	if err := writeReport(report, opts.Output); err != nil {
		return nil, fmt.Errorf("writing %s: %w", opts.Output, err)
	}
	return &report, nil
}

// Refresh rewrites upgrades.json after a local change (component toggle),
// from the cached manifest, fetching only when none is cached for the
// target tag. Nil when nothing was written.
func Refresh(opts CheckOptions) *Report {
	report, err := check(opts, true)
	if err != nil {
		slog.Warn("upgrades.json not refreshed", "err", err)
		return nil
	}
	return report
}

// RunCheck compares state.json against the target manifest and rewrites
// upgrades.json. Exit 0 up to date, 1 upgrade available, 2 error.
func RunCheck(stdout, stderr io.Writer, opts CheckOptions) int {
	report, err := check(opts, false)
	if err != nil {
		fmt.Fprintf(stderr, "Error: %v\n", err)
		return 2
	}
	printCheckSummary(stdout, *report)
	if report.UpgradeAvailable {
		return 1
	}
	return 0
}
